File: Libraries/EntityRegistry/main/mixed/io/denery/entityregistry/entity/CustomizableEntity.java
package io.denery.entityregistry.entity;

import com.ticxo.modelengine.api.ModelEngineAPI;
import com.ticxo.modelengine.api.model.ActiveModel;
import com.ticxo.modelengine.api.model.ModeledEntity;
import io.denery.entityregistry.EntityTypeRegistry;
import io.denery.entityregistry.behaviour.BehaviourCategories;
import org.bukkit.entity.LivingEntity;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.Optional;

/**
 * Represents the "actual" Customized Entity on a server.
 * @param <T> Type of living entity which will be "customized" with model and pathfinder.
 */
public class CustomizableEntity<T extends LivingEntity> {
    private static final Logger logger = LoggerFactory.getLogger("EntityRegistryLib");

    protected AbstractCustomizableEntityType type;
    protected T origin;
    private ModeledEntity modeledEntity;
    //protected BehaviourCategories.Behaviour behaviour;

    /*public CustomizableEntity(AbstractCustomizableEntityType type, T entity, BehaviourCategories behaviour) {
        this.type = type;
        this.origin = entity;
        //this.behaviour = behaviour;
    }

     */

    public CustomizableEntity(AbstractCustomizableEntityType type, T entity) {
        this.type = type;
        this.origin = entity;
    }

    /*public CustomizableEntity(String id, T entity, BehaviourCategories behaviour) {
        this.type = EntityTypeRegistry.getInstance().getRegisteredEntities().get(id);
        this.origin = entity;
        //this.behaviour = behaviour;
    }

     */

    public CustomizableEntity(String id, T entity) {
        this.type = EntityTypeRegistry.getInstance().getRegisteredEntities().get(id);
        this.origin = entity;
    }

    // Created For CustomizableEntityBuilder.
    protected CustomizableEntity() {}

    public void modelEntity() {
        ActiveModel activeModel = ModelEngineAPI.api.getModelManager().createActiveModel(type.id);
        Optional<ModeledEntity> me = Optional.ofNullable(ModelEngineAPI.api.getModelManager().createModeledEntity(origin));
        if (me.isEmpty()) throw new RuntimeException("Couldn't create modeled entity: " + origin.getName());
        modeledEntity = me.get();

        modeledEntity.addActiveModel(activeModel);
        modeledEntity.detectPlayers();
        modeledEntity.setInvisible(true);
    }

    public AbstractCustomizableEntityType getType() {
        return type;
    }

    public Optional<T> getOriginEntity() {
        return Optional.ofNullable(origin);
    }

    public Optional<ModeledEntity> getModeledEntity() {
        return Optional.ofNullable(modeledEntity);
    }

    //public Optional<BehaviourCategories.Behaviour> getBehaviour() {
    //    return Optional.ofNullable(behaviour);
    //}

    //public void setBehaviour(BehaviourCategories.Behaviour behaviour) {
    //    this.behaviour = behaviour;
    //}
}

File: Libraries/EntityRegistry/main/mixed/io/denery/entityregistry/entity/CustomizableEntity.scala
package io.denery.entityregistry.entity

import com.ticxo.modelengine.api.ModelEngineAPI
import com.ticxo.modelengine.api.model.{ActiveModel, ModeledEntity}
import io.denery.entityregistry.EntityTypeRegistry
import io.denery.entityregistry.entity.CustomizableEntityType
import io.denery.entityregistry.behaviour.BehaviourCategories.CommonBehaviourDictionary
import org.bukkit.entity.LivingEntity
import org.slf4j.{Logger, LoggerFactory}

import java.util.Optional

/**
 * Class which instance represents entity existing in the world that you can highly customize with model,
 * Behaviour (AI) and much more...
 *
 * @param _entityType type of entity which you registered before.
 * @param _origin the origin entity which custom entity is based. (Ocelot, villager, etc.)
 * @tparam T origin entity's type. (maybe will be deleted later)
 */
class CustomizableEntity[T <: LivingEntity](_entityType: CustomizableEntityType,
                                            _origin: T):

  def modelEntity(): Unit = {
    val activeModel = ModelEngineAPI.api.getModelManager.createActiveModel(entityType.getID
      .getOrElse(throw RuntimeException("Please, set entity's ID to find model engine's model!")))
    val me = Optional.ofNullable(ModelEngineAPI.api.getModelManager.createModeledEntity(origin))
    if (me.isEmpty) throw new RuntimeException("Couldn't create modeled entity: " + origin.getName)
    val modeledEntity: ModeledEntity = me.get
    modeledEntity.addActiveModel(activeModel)
    modeledEntity.detectPlayers()
    modeledEntity.setInvisible(true)
  }

  def performBehaviour(): Unit = {
    //TODO: Perform behaviour using BehaviourPerformerCategories class with Fibers.
  }

  def entityType: CustomizableEntityType = _entityType

  def origin: T = _origin

object CustomizableEntity:
  private val logger = LoggerFactory.getLogger("EntityRegistryLib")